vivisect/impemu/platarch/arm.py

```diff
@@ -1,14 +1,292 @@
         v_i_emulator.WorkspaceEmulator.__init__(self, vw, logwrite=logwrite, logread=logread)
+        self.setMemArchitecture(envi.ARCH_ARMV7)
+
+    def setThumbMode(self, thumb=1):
+        self.opcache = {}
+        e_arm.ArmEmulator.setThumbMode(self, thumb)
+
+    def setArmMode(self, arm=1):
+        self.opcache = {}
+        e_arm.ArmEmulator.setArmMode(self, arm)
+
+    def parseOpcode(self, va, arch=envi.ARCH_DEFAULT):
+        '''
+        Caching version.  
+        
+        We can make an opcode *faster* with the workspace because of
+        getByteDef etc... use it.
+
+        Made for ARM, because envi.Emulator doesn't understand the Thumb flag
+        '''
+        op = self.opcache.get(va)
+
+        if op == None:
+            tmode = self.getFlag(PSR_T_bit)
+            if arch == envi.ARCH_DEFAULT:
+                arch = (envi.ARCH_ARMV7, envi.ARCH_THUMB)[tmode]
+
+            op = envi.archs.arm.emu.ArmEmulator.parseOpcode(self, va, arch=arch)
+            self.opcache[va] = op
+        return op
+
+    def stepi(self):
+        # NOTE: when we step, we *always* want to be stepping over calls
+        # (and possibly import emulate them)
+        starteip = self.getProgramCounter()
+
+        # parse out an opcode
+        tmode = self.getFlag(PSR_T_bit)
+        #logger.debug("tmode: %x", tmode)
+        op = self.parseOpcode(starteip | tmode)
+        if self.emumon:
+            self.emumon.prehook(self, op, starteip)
+
+        # Execute the opcode
+        self.executeOpcode(op)
+        vg_path.getNodeProp(self.curpath, 'valist').append(starteip)
+
+        endeip = self.getProgramCounter()
+
+        if self.emumon:
+            self.emumon.posthook(self, op, endeip)
+
+        if not self.checkCall(starteip, endeip, op):
+            self.checkBranches(starteip, endeip, op)
+
+    def _prep(self, funcva, tmode=None):
+        if tmode != None:
+            # we're forcing thumb or arm mode... update the flag
+            self.setFlag(PSR_T_bit, tmode)
+            logger.debug("funcva thumb==%d  (forced):  0x%x", tmode, funcva)
+
+        elif funcva & 3:
+            # if the va isn't 4-byte aligned, it's gotta be thumb
+            self.setFlag(PSR_T_bit, 1)
+            funcva &= -2
+            logger.debug("funcva is THUMB(addr):  0x%x", funcva)
+
+        else:
+            loc = self.vw.getLocation(funcva)
+            if loc != None:
+                # if we have a opcode location, use it's iflags to determine mode
+                lva, lsz, lt, lti = loc
+                if (lti & envi.ARCH_MASK) == envi.ARCH_THUMB:
+                    self.setFlag(PSR_T_bit, 1)
+                    logger.debug("funcva is THUMB(loc):  0x%x", funcva)
+                else:
+                    logger.debug("funcva is ARM(loc):  0x%x", funcva)
+
+            else:
+                # otherwise, let's use some heuristics to guess.
+                armthumb = 0
+                armop  = None
```

Comment:
those are the most frequent first instructions for function prologs.  nothing terribly scientific, but "push" is preferable, and i believe i may have had some functions that overlapped, with thumb decoding as one and arm decoding as the other.  just wanted to make sure that whichever architecture was "push" won, but in the absence of "push," "ldr" was sufficient.

Reply:
although i'm now looking at it wondering why this isn't looking for some sort of "stm"... and instead looking at opcode numbers/constants instead of string comparisons.

Reply:
PUSH and LDR were sufficient.  we changed STM for PUSH a while back.